client_plugin/drivers/vfile/vfile_driver.go

```diff
@@ -494,6 +494,112 @@ func (d *VolumeDriver) processMount(r volume.MountRequest) volume.Response {
 	return volume.Response{Mountpoint: mountpoint}
 }
 
+func (d *VolumeDriver) addVMToClientList(volName string, vmIP string) error {
+	log.Infof("Add VM[%s] to clientList for volume[%s]", volName, vmIP)
+	var entries []kvstore.KvPair
+	var writeEntries []kvstore.KvPair
+	var volRecord etcdops.VFileVolConnectivityData
+	keys := []string{
+		kvstore.VolPrefixInfo + volName,
+	}
+	entries, err := d.kvStore.ReadMetaData(keys)
+	if err != nil {
+		// Failed to fetch existing metadata on the volume
+		// Set volume state to error as we cannot
+		// proceed
+		log.Warningf("Failed to read volume metadata before adding VM to ClientList: %v",
+			err)
+		return err
+	}
+	err = json.Unmarshal([]byte(entries[0].Value), &volRecord)
+	if err != nil {
+		// Failed to unmarshal record from JSON
+		log.Warningf("Failed to unmarshal JSON for reading existing metadata: %v",
+			err)
+		return err
+	}
+	// Add vmIP to CLientList
+	// then marshal the data structure to JSON again.
+	volRecord.ClientList = append(volRecord.ClientList, vmIP)
+	byteRecord, err := json.Marshal(volRecord)
+	if err != nil {
+		// Failed to marshal record as JSON
+		log.Warningf("Failed to marshal JSON for writing metadata: %v",
+			err)
+		return err
+	}
+	writeEntries = append(writeEntries, kvstore.KvPair{
+		Key:   kvstore.VolPrefixInfo + volName,
+		Value: string(byteRecord)})
+
+	log.Infof("Adding VM[%s] to CLientList for %s", vmIP, volName)
+	err = d.kvStore.WriteMetaData(writeEntries)
```

Comment:
How are you going to address concurrent invocations of addVMToClientList method? Say invocation-1 is adding IP1 and invocation-2 is adding IP2, and these could be running on different VFile programs. There is a possibility that invocaiton-2 could be overriding what invocation-1 updated in etcd. Basically, updating the kvstore has to be an atomic operation.